Delegate Python ~ (invert operator) to Tensor.bitwise_not().

xuhdev opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#22326 Delegate Python ~ (invert operator) to Tensor.bitwise_not().**
* #22320 Add a bitwise NOT operator for integer and Boolean types (CUDA).
* #22283 Add a bitwise NOT operator for integer and Boolean types (CPU).

Close #20024, Close #22246, Close #22262

Related #22324

Differential Revision: [D16183577](https://our.internmc.facebook.com/intern/diff/D16183577)

xuhdev committed:
Delegate Python ~ (invert operator) to Tensor.bitwise_not().

Close #20024, Close #22246, Close #22262

Related #22324

xuhdev committed:
Update on "Delegate Python ~ (invert operator) to Tensor.bitwise_not()."

Delegate Python ~ (invert operator) to Tensor.bitwise_not().

Close #20024, Close #22246, Close #22262

Related #22324

gh-metadata: pytorch pytorch 22326 gh/xuhdev/8/head

xuhdev commented:
This commit changes the `~` operator from being a simple `1-self` to have a bitwise NOT functionality, because [it is usually understood so in Python](https://docs.python.org/3/reference/expressions.html#unary-arithmetic-and-bitwise-operations). Indeed, one of the complaints that this PR intends to close, #22246, understands this operator as a "bitwise" operator.

Previously, as discussed in #4082, `~` was kept as `1-self` because ByteTensor was playing double duty of uint8 and bool. Given that we have BoolTensor now, do you think it is time to make the switch? @apaszke @colesbury

apaszke commented on tools/autograd/templates/python_variable_methods.cpp:
```diff
@@ -285,14 +285,14 @@ static PyObject * THPVariable_index_scalar(PyObject* self, PyObject* args) {
   auto& self_ = reinterpret_cast<THPVariable*>(self)->cdata;
-  if (self_.scalar_type() != at::kByte) {
-    throw TypeError("~ (operator.invert) is only implemented on byte tensors");
+  if (!isIntegralType(self_.scalar_type()) && self_.scalar_type() != at::kBool) {
```
Why isn't bool an integral type? It's not different than `uint1`

xuhdev replied:
It's defined [here](https://github.com/pytorch/pytorch/blob/cca923c4811dbc00acfe28ac67643a90b723236c/c10/core/ScalarType.h#L214). It's probably worth another independent PR for discussing it...

xuhdev replied:
It looks like current usage and definition of `isIntegralType` are more of semantic sense (bool is technically integral, but semantically it is not used as an integer). I would lean toward keep it as it is now.

Examples:

- https://github.com/pytorch/pytorch/blob/cca923c4811dbc00acfe28ac67643a90b723236c/aten/src/ATen/native/cpu/BinaryOpsKernel.cpp#L43
- https://github.com/pytorch/pytorch/blob/cca923c4811dbc00acfe28ac67643a90b723236c/torch/csrc/autograd/python_variable_indexing.cpp#L171

apaszke commented on test/test_torch.py:
```diff
@@ -11246,9 +11242,18 @@ def test_bitwise_ops(self):
-        x = torch.ByteTensor([0, 1, 1])
-        self.assertEqual((~x).tolist(), [1, 0, 0])
+    def test_op_invert(self):
+        for t in (torch.ByteTensor, torch.ShortTensor, torch.IntTensor, torch.LongTensor):
```
It would be better to make this iterate over the dtypes (e.g. `(torch.byte, torch.short, torch.int, torch.long)`)

xuhdev replied:
Makes sense.

xuhdev committed:
Update on "Delegate Python ~ (invert operator) to Tensor.bitwise_not()."

Delegate Python ~ (invert operator) to Tensor.bitwise_not().

Close #20024, Close #22246, Close #22262

Related #22324

gh-metadata: pytorch pytorch 22326 gh/xuhdev/8/head

xuhdev committed:
Update on "Delegate Python ~ (invert operator) to Tensor.bitwise_not()."

Delegate Python ~ (invert operator) to Tensor.bitwise_not().

Close #20024, Close #22246, Close #22262

Related #22324

gh-metadata: pytorch pytorch 22326 gh/xuhdev/8/head

xuhdev commented:
@pytorchbot retest this please

xuhdev commented:
@pytorchbot retest this please

colesbury commented on test/test_torch.py:
```diff
@@ -11251,9 +11247,18 @@ def test_bitwise_ops(self):
-        x = torch.ByteTensor([0, 1, 1])
-        self.assertEqual((~x).tolist(), [1, 0, 0])
+    def test_op_invert(self):
+        for t in (torch.uint8, torch.int8, torch.int16, torch.int32, torch.int64):
```
You should test actual values. This only tests that ~ is the same as bitwise_not. i.e you should test that ~torch.tensor([True, False]) is [False, True]. Same with integer types.

colesbury commented on test/test_torch.py:
```diff
@@ -11251,9 +11247,18 @@ def test_bitwise_ops(self):
+        for t in (torch.uint8, torch.int8, torch.int16, torch.int32, torch.int64):
+            a = torch.arange(127, dtype=t)
+            self.assertEqual(~a, a.bitwise_not())
+        a = torch.BoolTensor([True, False])
```
Use `torch.tensor([True, False])` (dtype is inferred)

We are avoiding the classes like torch.BoolTensor and torch.FloatTensor in favor of dtypes

xuhdev committed:
Update on "Delegate Python ~ (invert operator) to Tensor.bitwise_not()."

Delegate Python ~ (invert operator) to Tensor.bitwise_not().

Close #20024, Close #22246, Close #22262

Related #22324

gh-metadata: pytorch pytorch 22326 gh/xuhdev/8/head

xuhdev committed:
Update on "Delegate Python ~ (invert operator) to Tensor.bitwise_not()."

Delegate Python ~ (invert operator) to Tensor.bitwise_not().

Close #20024, Close #22246, Close #22262

Related #22324

gh-metadata: pytorch pytorch 22326 gh/xuhdev/8/head

xuhdev committed:
Update on "Delegate Python ~ (invert operator) to Tensor.bitwise_not()."

Delegate Python ~ (invert operator) to Tensor.bitwise_not().

Close #20024, Close #22246, Close #22262

Related #22324

gh-metadata: pytorch pytorch 22326 gh/xuhdev/8/head

xuhdev committed:
Update on "Delegate Python ~ (invert operator) to Tensor.bitwise_not()."

Delegate Python ~ (invert operator) to Tensor.bitwise_not().

Close #20024, Close #22246, Close #22262

Related #22324

gh-metadata: pytorch pytorch 22326 gh/xuhdev/8/head

xuhdev committed:
Update on "Delegate Python ~ (invert operator) to Tensor.bitwise_not()."

Delegate Python ~ (invert operator) to Tensor.bitwise_not().

Close #20024, Close #22246, Close #22262

Related #22324

gh-metadata: pytorch pytorch 22326 gh/xuhdev/8/head

xuhdev committed:
Update on "Delegate Python ~ (invert operator) to Tensor.bitwise_not()."

Delegate Python ~ (invert operator) to Tensor.bitwise_not().

Close #20024, Close #22246, Close #22262

Related #22324

gh-metadata: pytorch pytorch 22326 gh/xuhdev/8/head

xuhdev committed:
Update on "Delegate Python ~ (invert operator) to Tensor.bitwise_not()."

Delegate Python ~ (invert operator) to Tensor.bitwise_not().

Close #20024, Close #22246, Close #22262

Related #22324

gh-metadata: pytorch pytorch 22326 gh/xuhdev/8/head

xuhdev committed:
Update on "Delegate Python ~ (invert operator) to Tensor.bitwise_not()."

Delegate Python ~ (invert operator) to Tensor.bitwise_not().

Close #20024, Close #22246, Close #22262

Related #22324

gh-metadata: pytorch pytorch 22326 gh/xuhdev/8/head

xuhdev committed:
Update on "Delegate Python ~ (invert operator) to Tensor.bitwise_not()."

Delegate Python ~ (invert operator) to Tensor.bitwise_not().

Close #20024, Close #22246, Close #22262

Related #22324

gh-metadata: pytorch pytorch 22326 gh/xuhdev/8/head

xuhdev committed:
Update on "Delegate Python ~ (invert operator) to Tensor.bitwise_not()."

Delegate Python ~ (invert operator) to Tensor.bitwise_not().

Close #20024, Close #22246, Close #22262

Related #22324

gh-metadata: pytorch pytorch 22326 gh/xuhdev/8/head

xuhdev committed:
Update on "Delegate Python ~ (invert operator) to Tensor.bitwise_not()."

Delegate Python ~ (invert operator) to Tensor.bitwise_not().

Close #20024, Close #22246, Close #22262

Related #22324

gh-metadata: pytorch pytorch 22326 gh/xuhdev/8/head

xuhdev committed:
Update on "Delegate Python ~ (invert operator) to Tensor.bitwise_not()."

Delegate Python ~ (invert operator) to Tensor.bitwise_not().

Close #20024, Close #22246, Close #22262

Related #22324

gh-metadata: pytorch pytorch 22326 gh/xuhdev/8/head

xuhdev committed:
Update on "Delegate Python ~ (invert operator) to Tensor.bitwise_not()."

Delegate Python ~ (invert operator) to Tensor.bitwise_not().

Close #20024, Close #22246, Close #22262

Related #22324

gh-metadata: pytorch pytorch 22326 gh/xuhdev/8/head

xuhdev committed:
Update on "Delegate Python ~ (invert operator) to Tensor.bitwise_not()."

Delegate Python ~ (invert operator) to Tensor.bitwise_not().

Close #20024, Close #22246, Close #22262

Related #22324

gh-metadata: pytorch pytorch 22326 gh/xuhdev/8/head

colesbury commented:
@xuhdev is this stack of PRs ready?

xuhdev commented:
@colesbury Yes.

xuhdev commented:
@pytorchbot merge this please

colesbury commented:
I'm reverting this for now. The problem is that comparison operators still return torch.uint8 and now `~` behaves differently. Once comparison operators return torch.bool we can re-land this
